dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs

```diff
@@ -0,0 +1,316 @@
+    /// <returns>Search client ready to read/write</returns>
+    private async Task<SearchClient> GetSearchClientAsync(
+        string indexName,
+        CancellationToken cancellationToken = default)
+    {
+        Response<SearchIndex>? existingIndex = null;
+        try
+        {
+            // Search the index
+            existingIndex = await this._adminClient.GetIndexAsync(indexName, cancellationToken).ConfigureAwait(false);
+        }
+        catch (RequestFailedException e) when (e.Status == 404)
+        {
+        }
+
+        // Create the index if it doesn't exist
+        if (existingIndex == null || existingIndex.Value == null)
+        {
+            await this.CreateIndexAsync(indexName, cancellationToken).ConfigureAwait(false);
+        }
+
+        // Search an available client from the local cache
+        if (!this._clientsByIndex.TryGetValue(indexName, out SearchClient? client) || client == null)
+        {
+            client = this._adminClient.GetSearchClient(indexName);
+            this._clientsByIndex[indexName] = client;
+        }
+
+        return client;
+    }
+
+    /// <summary>
+    /// Create a new search index.
+    /// </summary>
+    /// <param name="indexName">Index name</param>
+    /// <param name="cancellationToken">Task cancellation token</param>
+    private Task CreateIndexAsync(
+        string indexName,
+        CancellationToken cancellationToken = default)
+    {
+        var fieldBuilder = new FieldBuilder();
+        var fields = fieldBuilder.Build(typeof(AzureCognitiveSearchRecord));
+        var newIndex = new SearchIndex(indexName, fields)
+        {
+            SemanticSettings = new SemanticSettings
+            {
+                Configurations =
+                {
+                    new SemanticConfiguration("default", new PrioritizedFields
+                    {
+                        TitleField = new SemanticField { FieldName = "Description" },
+                        ContentFields =
+                        {
+                            new SemanticField { FieldName = "Text" },
+                            new SemanticField { FieldName = "AdditionalMetadata" },
+                        }
+                    })
+                }
+            }
+        };
+
+        return this._adminClient.CreateIndexAsync(newIndex, cancellationToken);
+    }
+
+    private async Task<string> UpsertRecordAsync(
+        string indexName,
+        AzureCognitiveSearchRecord record,
+        CancellationToken cancellationToken = default)
+    {
+        var client = await this.GetSearchClientAsync(indexName, cancellationToken).ConfigureAwait(false);
+
+        Response<IndexDocumentsResult>? result = await client.MergeOrUploadDocumentsAsync(
+            new List<AzureCognitiveSearchRecord> { record },
+            new IndexDocumentsOptions { ThrowOnAnyError = true },
+            cancellationToken).ConfigureAwait(false);
+
+        if (result == null)
```

Comment:
Should we check `result.Value.Results.Count > 0` as well?

Reply:
it's very unlikely, but I added the check